Refactor Node.js code to ESM

cb1kenobi opened the pull request:
This is a major refactor that converts all JavaScript code from CommonJS to ESM. This affects all CLI commands (create, clean, build, and project) and the scons tasks.

This PR does NOT update npm dependencies. The dependencies are horribly behind, but this PR is already huge.

Additionally, all i18n related stuff in the CLI has been removed. We never added strings for all the various languages anyways.

I remove the `scons-check-lockfile.js` file and removed it from the lint-staged config. Somewhere, npm changed the `package-lock.json` format and it's totally incorrect now. I'm not sure why we needed it in the first place and I don't think we need it anymore.

This will require Titanium CLI v7 or newer and Node.js 18 or newer. Titanium CLI v7 was released May 10, 2024 and we have since shipped CLI v8. Titanium SDK 13 (master branch) dropped Node.js 16 and older. This means this PR is no longer a breaking change.

Note: This PR has been tested on Windows 11 + Node.js 24 + Titanium CLI 8.1.0.

cb1kenobi committed:
refactor: migrate cli dir to esm

cb1kenobi committed:
chore: finished migrating android code to esm

cb1kenobi committed:
Merge branch 'master' into esm

cb1kenobi committed:
chore: migrate build dir to esm

cb1kenobi committed:
chore: start refactoring ios files

cb1kenobi committed:
chore: refactored iOS build

cb1kenobi committed:
chore: more work on the esm migration

cb1kenobi committed:
chore: fixing esm stuff

cb1kenobi committed:
chore: fix merge conflicts

cb1kenobi committed:
Merge branch 'master' into esm

cb1kenobi committed:
Merge branch 'master' into esm

cb1kenobi committed:
Merge branch 'master' into esm

cb1kenobi committed:
chore: fix merge conflict

cb1kenobi committed:
Merge branch 'master' into esm

cb1kenobi committed:
chore: fix merge conflicts

cb1kenobi committed:
chore: fix merge conflicts

cb1kenobi committed:
chore: fix merge

cb1kenobi committed:
chore: fix core-js detection

cb1kenobi committed:
chore: fix bugs

cb1kenobi committed:
chore: fix ios sdk build

cb1kenobi committed:
chore: debug ios package

cb1kenobi committed:
chore: more debug ios package

cb1kenobi committed:
chore: use execFileSync

cb1kenobi committed:
chore: revert execFileSync

cb1kenobi committed:
chore: promisify

cb1kenobi committed:
chore: fix itunes artwork merge

m1ga commented on android/cli/hooks/run.js:
```diff
@@ -156,61 +152,61 @@ exports.init = function (logger, config, cli) {
 										err.split('\n').forEach(logger.error);
 										if (err.indexOf('INSTALL_PARSE_FAILED_NO_CERTIFICATES') !== -1) {
-											logger.error(__('Make sure your keystore is signed with a compatible signature algorithm such as "SHA1withRSA" or "MD5withRSA".'));
+											logger.error('Make sure your keystore is signed with a compatible signature algorithm such as "SHA1withRSA" or "MD5withRSA".');
```
I think MD5 is not recommended anymore but I can't find any info about it at the moment, so we can keep it unless I find something else

m1ga commented on build/lib/packager.js:
```diff
@@ -32,6 +38,8 @@ const TITANIUM_PREP_LOCATIONS = [
 async function zip(cwd, filename) {
 	const command = os.platform() === 'win32' ? path.join(ROOT_DIR, 'build/win32/zip') : 'zip';
 	const params = os.platform() === 'win32' ? '-9 -q -r' : '-9 -q -r -y';
+
+	console.log(`exec: ${command} ${params} "${path.join('..', path.basename(filename))}" *`);
```
```suggestion
```

think that was just for debugging, can be removed again

m1ga commented:
Some app/module build errors using this SDK:

hyperloop-example repo:
```
[INFO]  Writing bootstrap json
[ERROR] TypeError: Cloak is not a constructor
    at AndroidBuilder.encryptJSFiles (file:///home/miga/.titanium/mobilesdk/linux/13.1.0/android/cli/commands/_build.js:2863:34)
    at async Promise.all (index 0)
    at async AndroidBuilder.generateAppProject (file:///home/miga/.titanium/mobilesdk/linux/13.1.0/android/cli/commands/_build.js:2313:3)
    at async AndroidBuilder.run (file:///home/miga/.titanium/mobilesdk/linux/13.1.0/android/cli/commands/_build.js:1624:4)
```

kitchensink repo:
```
[INFO]  Found Titanium plugin id=ti.alloy version=1.0 
TypeError: Cannot read properties of undefined (reading 'devDependencies')
    at getWebpackProjectType (file:///home/miga/.titanium/mobilesdk/linux/13.1.0/cli/hooks/webpack.js:137:34)
    at async file:///home/miga/.titanium/mobilesdk/linux/13.1.0/cli/hooks/webpack.js:45:17
    at async file:///usr/local/lib/node_modules/titanium/src/cli.js:346:8
```

android module build:
```
Command
  /usr/local/bin/node /usr/local/bin/ti build -p android --build-only --sdk 13.1.0

[LiveView] ReferenceError: module is not defined
    at file:///home/miga/.titanium/mobilesdk/linux/13.1.0/android/cli/commands/_buildModule.js:958:48
    at ModuleJobSync.runSync (node:internal/modules/esm/module_job:395:35)
    at ModuleLoader.importSyncForRequire (node:internal/modules/esm/loader:329:47)
    at loadESMFromCJS (node:internal/modules/cjs/loader:1411:24)
    at Module._compile (node:internal/modules/cjs/loader:1544:5)
    at Object..js (node:internal/modules/cjs/loader:1668:16)
    at Module.load (node:internal/modules/cjs/loader:1313:32)
    at Function._load (node:internal/modules/cjs/loader:1123:12)
    at TracingChannel.traceSync (node:diagnostics_channel:322:14)
    at wrapModuleLoad (node:internal/modules/cjs/loader:217:24)
Warning: Detected unsettled top-level await at file:///usr/local/lib/node_modules/titanium/src/main.js:8
        await cli.go();
        ^
```
(not using liveview)

m1ga reviewed (commented):
you can remove `node-uuid` from the package.json and use crypto.randomUUID() (https://developer.mozilla.org/en-US/docs/Web/API/Crypto/randomUUID)

cb1kenobi committed:
Update build/lib/packager.js

Co-authored-by: Michael Gangolf <m1ga@users.noreply.github.com>

cb1kenobi committed:
chore: fix pr feedback

cb1kenobi commented on cli/lib/creators/module.js:
```diff
@@ -10,308 +10,284 @@
+			const variables = {
+					author: this.config.get('user.name', 'Your Name'),
+					publisher: this.config.get('app.publisher', 'Your Company'),
+					guid: uuid.v4(),
```
Fixed!

cb1kenobi commented on cli/lib/creators/app.js:
```diff
@@ -10,219 +10,194 @@
+import fs from 'fs-extra';
+import path from 'node:path';
+import ti from 'node-titanium-sdk';
+import uuid from 'node-uuid';
```
Fixed!

cb1kenobi committed:
chore: fix android module build bug

cb1kenobi commented:
I have fixed all but one of the bugs you have found.

The create project bug `TypeError: Cannot read properties of null (reading 'cyan')` is happening because `cli.argv.type` is `undefined`. When you run `ti create`, the CLI sees that `--type` is not specified, so it correctly gets the `type` prompt and then invokes it: https://github.com/tidev/titanium-cli/blob/main/src/cli.js#L1178-L1195. However, the CLI doesn't actually prompt! I have no idea why yet. I need to investigate more.

Two other interesting things I noticed:
1. When I launch KitchenSink v2 in the Android emulator, it loads, but shows a error that I can skip or quit the app. Skipping causes the app to essentially break.
2. CI build/iOS is now failing due to `xcrun: error: missing DEVELOPER_DIR path: /Applications/Xcode_26.0.0.app/Contents/Developer`. Hopefully this just a fluke and goes away with the next push.

cb1kenobi committed:
chore: fix markdown import

cb1kenobi committed:
chore: fix spawn import

cb1kenobi committed:
chore: fix project creation

cb1kenobi committed:
chore: use defaults

cb1kenobi committed:
chore: use defaults again

cb1kenobi committed:
chore: no xcode version

cb1kenobi committed:
chore: debug build titaniumkit

cb1kenobi committed:
chore: add delay

cb1kenobi committed:
chore: remove patch if minor and patch are zero

cb1kenobi committed:
chore: remove debug

cb1kenobi committed:
chore: fix lint

cb1kenobi committed:
chore: remove unused import

cb1kenobi commented:
@m1ga I've fixed all of the issues you found. I added a workaround for GitHub's bad xcode path and fixed a couple JS lint errors.

I've done a bunch of testing with creating, building, and cleaning projects and things look good.

cb1kenobi committed:
chore: remove another unused import

m1ga commented:
Awesome. Linux tests are all good now (sdk, module create/build, app create/build) 👍 

iOS build is broken:
```
% ti build -p ios --sdk 13.1.0
Titanium Command-Line Interface v8.1.0 SDK v13.1.0
Copyright TiDev, Inc. 4/7/2022-Present. All Rights Reserved.

Want to help? https://tidev.io/donate or https://tidev.io/contribute

[ERROR] Invalid platform "undefined"

Available platforms for SDK version 13.1.0:
    ios
```

also tested it without -p and --platform. Same errors

cb1kenobi committed:
Merge branch 'master' into esm

cb1kenobi committed:
chore: restore original xcode path resolution

cb1kenobi committed:
chore: fix colors import

cb1kenobi committed:
chore: fix png import

cb1kenobi committed:
chore: fix ios module build

m1ga commented:
still the same error here, it's just line 32 now :) 
`.../iphone/cli/commands/_buildModule.js:32:45`

`ti build -p ios -b --sdk 13.1.0` with the ti.sslcheck module.
Will do some more tests tomorrow

cb1kenobi committed:
chore: remove module

m1ga commented:
Building a module works fine now with the latest update. 
Did you change the SDK version in titanium.xconfig? If not: that could already be the error.

iOS error messages are not very informative :) See: https://github.com/tidev/titanium-cli/issues/593 : it will show you the list of all simulators if you just have a `missing the private key` error that is visible in the trace log. It looks like Hendrik is trying to improve those: https://github.com/tidev/titanium-sdk/issues/14290

cb1kenobi committed:
chore: scope node imports

cb1kenobi committed:
chore: remove useless await

cb1kenobi committed:
chore: fix import

cb1kenobi committed:
Merge branch 'master' into esm

cb1kenobi committed:
chore: fix test to use latest Titanium CLI

m1ga reviewed (approved):
LGTM :rocket:
